### PR TITLE
[WIP] fast random projection

### PR DESCRIPTION
**Pull request by ghost**
TLDR: [Fast Random Projection](https://arxiv.org/abs/1908.11512) implemented. `HR@5 37014 55187 0.670701433308569`
 `HR@10 46608 55187 0.8445467229601175`. Very good bang for buck in terms of speed:quality tradeoff. Run `fast_cloverleaf.py` in the examples directory to try it.

I'm opening this early PR so you can take `FastRP` for a spin. There's a few finishing touches to add before promoting to full PR. Now is a good time to share any feedback/thoughts on the interface / changes. Highly recommend running this on a large-ish dataset and seeing how much RAM it soaks up.

Some "plumbing" changes that were necessary: exposing the underlying `CSR` data within the crate, so one can perform matrix operations on it. Edge-deduplication was introduced, but this can be made optional. Debate is welcome.

Peak memory usage should be exactly  `2 * N * dim * 4` bytes at any given time, this is due to needing to add power `k-1` and power `k` of the transition matrix at each iteration.

TODOs:

0) normalization strength `Beta` should be added as part of the initial "normalization" step. I'm using `1` for now. This is one more hyper parameter.
1) Temporarily this uses twice as much memory as necessary, need to implement Sparse-Sparse matrix multiply using borrowed data, and that should cut it down by half.
2) Borrowing some ideas from another method `ProNE` can improve quality, as I mentioned in the [prior PR](  https://github.com/Refefer/cloverleaf/pull/8).

### CODE REVIEW

File: src/algos/fastrp.rs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    weights: &[f32],
+    norm_powers: bool,
+    rng: &mut impl Rng
+) -> DMatrix<f32> {
+
+    let k = weights.len();
+
+    // Compute normalizer matrix (inverse degree matrix)
+    let mut a_sum = csr_row_sum(&A);
+    a_sum.iter_mut().for_each(|x| *x = 1.0 / *x); // FIXME: there is a possibility for division by zero here
```
Comment (ghost) [inline src/algos/fastrp.rs]: @Refefer wasn't sure how to deal with this. In the case that a row of the matrix sums to zero, this will lead to division by zero.
